src/compute/src/logging/compute.rs

```diff
@@ -609,32 +609,33 @@ where
         import_id: GlobalId,
         export_ids: Vec<GlobalId>,
     ) -> Self {
-        let mut previous_time = None;
+        // Prepare retraction events as `LogOnDrop`s, to ensure they are logged even when the
+        // dataflow is dropped before the input advances to the empty frontier.
+        let mut retractions = Vec::<LogOnDrop>::with_capacity(export_ids.len());
+
         self.inspect_container(move |event| {
-            if let Err(frontier) = event {
-                if let Some(previous) = previous_time {
-                    for &export_id in export_ids.iter() {
-                        logger.log(ComputeEvent::ImportFrontier {
-                            import_id,
-                            export_id,
-                            time: previous,
-                            diff: -1,
-                        });
-                    }
-                }
-                if let Some(time) = frontier.get(0) {
-                    for &export_id in export_ids.iter() {
-                        logger.log(ComputeEvent::ImportFrontier {
-                            import_id,
-                            export_id,
-                            time: *time,
-                            diff: 1,
-                        });
-                    }
-                    previous_time = Some(*time);
-                } else {
-                    previous_time = None;
-                }
+            let Err(frontier) = event else { return };
+
+            // Log frontier retractions by dropping them.
+            retractions.drain(..).for_each(drop);
```

Comment:
Is this different than `retractions.clear()`?

Reply:
No, it's just more explicit. But I guess with the comment the explicitly is not really needed.